### DISCUSSION TITLE
Read the terminal text

### DISCUSSION THREAD

**Discussion by dhananjayharel** [eclipse-theia/theia#9662] [Q&A]
### Description

Need to read the text in the terminal

### Requirement
I am firing some command using terminal's send text api, then I need to read the output of that
command after complete but I havent found any method or API yet.

**Comment by akosyakov** [comment]
I don't think it is possible. One has to use tasks or processes to get an output. Terminal are just character streams, there are not such high-level concepts like a command or a task.

cc @theia-ide/task-extension correct me if i'm wrong, also suggest an alternative how to use a task or process extension for it

**Comment by dhananjayharel** [comment]
Thanks .. I think process extension will be the best option for my requirement but need some details on it, ie how to use it . very basic example  would also work.

**Comment by akosyakov** [comment]
@marechal-p maybe you can give a hint

**Comment by dhananjayharel** [comment]
@marechal-p 
For example: I want to add a menu option in the editor, on click of it should execute the command and read the output of that command.
I have seen the theia@process code files but I couldn't able to find any front end API/ method to call the backend process.

**Comment by akosyakov** [comment]
You can also define your own backend service and use nodejs directly. It is not clear to me still when process extension should be used and when not. I know one case for cloud scenarios, that some processes should be run by a user operating system and some by other container for example. In this case I would say that process extension should be used for user processes and nodejs directly to run in another container.

**Comment by paul-marechal** [comment]
TBF I would really like to refactor the process extension a bit, and remove the spawning from inside the constructors, into factory functions. In a nutshell, the process extension normalizes how nodejs and node-pty processes behave and how to consume their output from different clients (using multi-reader ring-buffers for instance). Could not find the exact place where to look for, but either make your own backend service, or reuse the `TerminalServer` one, if that fits.

I'm on vacation for the time being, I'll get back to you if you still have issues around end of next week.

edit: I had other priorities.
